distributor/lru.go

```diff
@@ -32,8 +32,12 @@ func (lru *cache) Put(key string, value interface{}) {
 	}
 	lru.mut.Lock()
 	defer lru.mut.Unlock()
-	if _, ok := lru.get(key); ok {
-		return
+	if v, ok := lru.get(key); ok {
+		if v == value {
+			return
+		} else {
+			lru.priority.Remove(lru.priority.Front())
```

Comment:
Ok, I understand now.  So it is not wrong, but you are using a side-effect of the get() function.  If at some future time get() does not do that, this code would remove a different element.

Instead, find the element, and remove it.  That will always work.